src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+                        }
+                        Ok((inner_key, AccountSignaturesVerificationEntry {
+                            signature:  signature.clone(),
+                            public_key: public_key.clone(),
+                        }))
+                    })
+                    .collect();
+
+                outer_map.insert(outer_key, CredentialSignaturesVerificationData {
+                    data: inner_map?,
+                });
+            } else {
+                return Err(SignatureError::MismatchMapIndices);
+            }
+        }
+
+        Ok(AccountSignaturesVerificationData { data: outer_map })
+    }
+}
+
+/// Calculate the message hash that is signed in Concordium wallets.
+pub fn calculate_message_hash(message: &Message<'_>, signer: AccountAddress) -> [u8; 32] {
+    let message_signed_in_wallet = match message {
+        Message::BinaryMessage(message) => message,
+        Message::TextMessage(message) => (**message).as_bytes(),
+    };
+
+    // A message signed in a Concordium wallet is prepended with the
+    // `account` address (signer) and 8 zero bytes. Accounts in a Concordium
+    // wallet can either sign a regular transaction (in that case the
+    // prepend is `account` address and the nonce of the account which is by
+    // design >= 1) or sign a message (in that case the prepend is `account`
+    // address and 8 zero bytes). Hence, the 8 zero bytes ensure that the user
+    // does not accidentally sign a transaction. The account nonce is of type
+    // u64 (8 bytes).
+    sha2::Sha256::digest(
+        [
+            &signer.as_ref() as &[u8],
+            &[0u8; 8],
+            &message_signed_in_wallet,
+        ]
+        .concat(),
+    )
+    .into()
+}
+
+/// Check that all key indices in the signatures map exist on chain but not vice
+/// versa.
+fn assert_signature_map_key_indices_on_chain<C: Curve, AttributeType: Attribute<C::Scalar>>(
+    signatures: &AccountSignatures,
+    on_chain_credentials: &BTreeMap<
+        CredentialIndex,
+        Versioned<AccountCredentialWithoutProofs<C, AttributeType>>,
+    >,
+) -> Result<(), SignatureError> {
+    // Ensure all outer-level keys in the signatures map exist in the
+    // on_chain_credentials map.
+    for outer_key in signatures.sigs.keys() {
+        // Check if the outer_key exists in the on_chain_credentials map.
+        let on_chain_cred = on_chain_credentials
+            .get(&CredentialIndex { index: *outer_key })
+            .ok_or(SignatureError::MissingIndicesOnChain {
+                credential_index: *outer_key,
+                // The key_index does not exist in this context, use the default value.
+                key_index:        0u8,
+            })?;
+
+        // Ensure that the inner-level keys in the signatures map exist in the
+        // on_chain_credentials map.
+        for inner_key in signatures.sigs[outer_key].sigs.keys() {
```

Comment:
**Suggestion (non-blocking):** Really minor thing, but we can avoid a lookup in the inner map:

```suggestion
    for (outer_key, inner_map) in signatures.sigs {
        // Check if the outer_key exists in the on_chain_credentials map.
        let on_chain_cred = on_chain_credentials
            .get(&CredentialIndex { index: *outer_key })
            .ok_or(SignatureError::MissingIndicesOnChain {
                credential_index: *outer_key,
                // The key_index does not exist in this context, use the default value.
                key_index:        0u8,
            })?;

        // Ensure that the inner-level keys in the signatures map exist in the
        // on_chain_credentials map.
        for inner_key in inner_map.sigs.keys() {
```